google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -181,6 +181,258 @@ class ResumableAsyncStreamingReadWriteRpc {
+          stream_state_ = State::kShutdown;
+          if (in_progress_read_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_read_->get_future()));
+          }
+          if (in_progress_write_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_write_->get_future()));
+          }
+          std::shared_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+              stream = std::move(stream_);
+          CompleteStreamLockHeld(Status());
+          root_future = root_future.then([stream](future<void>) {
+            return future<void>(stream->Finish());
+          });
+      }
+    }
+    root_promise.set_value();
+    return root_future;
+  }
+
+ private:
+  using UnderlyingStream =
+      std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>;
+
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  void SetReadWriteFuturesLockHeld() {
+    absl::optional<promise<void>> read_reinit_done;
+    absl::optional<promise<void>> write_reinit_done;
+
+    {
+      read_reinit_done.swap(read_reinit_done_);
+      write_reinit_done.swap(write_reinit_done_);
+    }
+    mu_.unlock();
+    if (read_reinit_done) read_reinit_done->set_value();
+
+    if (write_reinit_done) write_reinit_done->set_value();
+    mu_.lock();
+  }
+
+  void CompleteStreamLockHeld(Status status) {
+    SetReadWriteFuturesLockHeld();
+    status_promise_.set_value(std::move(status));
+  }
+
+  void FinishRetryPromiseLockHeld() {
+    assert(retry_promise_.has_value());
+    promise<void> retry_promise = std::move(*retry_promise_);
+    retry_promise_.reset();
+    mu_.unlock();
+    retry_promise.set_value();
+    mu_.lock();
+  }
+
+  void AttemptRetry(Status const& status,
+                    std::shared_ptr<RetryPolicy> retry_policy,
+                    std::shared_ptr<BackoffPolicy> backoff_policy) {
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ == State::kShutdown) {
+        return FinishRetryPromiseLockHeld();
+      }
+      assert(stream_state_ == State::kRetrying);
+    }
+
+    if (!retry_policy->IsExhausted() && retry_policy->OnFailure(status)) {
+      sleeper_(backoff_policy->OnCompletion())
+          .then([this, retry_policy, backoff_policy](future<void>) {
+            Initialize(retry_policy, backoff_policy);
+          });
+      return;
+    }
+
+    std::lock_guard<std::mutex> g{mu_};
+    FinishRetryPromiseLockHeld();
+    if (stream_state_ == State::kShutdown) return;
+    stream_state_ = State::kShutdown;
+
+    CompleteStreamLockHeld(status);
+  }
+
+  void Initialize(std::shared_ptr<RetryPolicy> retry_policy,
+                  std::shared_ptr<BackoffPolicy> backoff_policy) {
+    future<bool> start_future;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ == State::kShutdown)
+        return FinishRetryPromiseLockHeld();
+      stream_ = stream_factory_();
```

Comment:
decided to use `std::shared_ptr<std::unique_ptr<StreamType>>` to facilitate only initializing `stream_` on success
otherwise `stream_` would have to be a `std::shared_ptr` which we didn't think made much sense

Reply:
I don't follow: `std::shared_ptr<StreamType>` does not make sense, but `std::shared_ptr<std::unique_ptr<StreamType>>` makes sense?  Please don't do that.

Reply:
the reason being that upon successful initialization, it only makes sense for the class to have ownership of the underlying stream
though it sounds like the weirdness of the double indirection via `std::shared_ptr<std::unique_ptr<StreamType>>` outweighs the reasoning behind keeping the member variable as an `std::unique_ptr`?

Reply:
> though it sounds like the weirdness of the double indirection via `std::shared_ptr<std::unique_ptr<StreamType>>` outweighs the reasoning behind keeping the member variable as an `std::unique_ptr`?

Yes.  I think the recommendation to prefer `std::unique_ptr<>` over `std::shared_ptr<>` is basically sound, but I do not see it as a suicide pact.  When the complexity of using `std::unique_ptr<>` gets too high I just shrug and switch to `std::shared_ptr<>`.

Reply:
I think in particular, shared_ptrs to streams are a bit scary, since they have some fairly tight constraints on calling their methods (specifically, that there can be exactly one future outstanding to read/write at a time).

Do you think what is done here adds significant complexity over the gain from knowing we're the only owner? If not, I'd rather leave unique_ptrs in the interface and members.

Reply:
AFAIK, the only thing preventing two outstanding `Read()` calls (or `Write()` calls) is a comment in the header:

https://github.com/googleapis/google-cloud-cpp/blob/faecca1324a6c42af2a4f5fdd5018c9ebb9bea73/google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h#L122-L124

 How is the `std::unique_ptr<>` helping more or less with that?  And furthermore, how would this `std::shared_ptr<>` leave the confines of this class and become reachable from multiple points?

Admittedly the use of `std::shared_ptr<std::unique_ptr<>>` is also fairly constrained, but I will not understand the code or motivation by Monday.  Forget about 6 months or 2 years from now.

I think we have enough complexity to carry in our heads, cutting one level of indirection seems like a good thing to me.

Reply:
In particular, the initializer is required to return ownership of the unique_ptr to the caller, at which point, if it has outstanding futures, it can no longer enforce their contracts, so it must not have any outstanding futures. 

If it's a shared_pointer, one could quite easily store a reference to the underlying stream after initialization is complete, and think that is a perfectly fine thing to do.

I don't feel very strongly about this, other than unique_ptr is the more constrained API that gives us greater confidence that we are the only valid owner, which shared_ptr doesn't let you reason about. I think from a code archaeology perspective, down the road when modifying this, if I see a shared_ptr member, I assume someone else must be potentially accessing it, which it is confusing that in this case that is not true. 

Given the above, has your opinion changed at all on this? If not, we can make this change.

Reply:
Hmm... I think you are saying that `std::shared_ptr<std::unique_ptr<>>` can be reasoned about locally, as it is used in a single function, while changing the initializer to return `std::shared_ptr<>` forces us to reason about something across multiple functions.  Fair enough.  That deserves at least a comment?

Finally, can we remove those temporary futures.  Chaining `.then()` calls seems more readable (module the formatting robots doing a semi-decent job).

Reply:
yup, just pushed